pykokkos/lib/ufuncs.py

```diff
@@ -256,3 +256,491 @@ def sign(view):
+@pk.workunit
+def matmul_impl_1d_double(tid: int, acc: pk.Acc[pk.double], viewA: pk.View1D[pk.double], viewB: pk.View2D[pk.double]):
+    acc += viewA[tid] * viewB[0][tid]
+
+@pk.workunit
+def matmul_impl_1d_float(tid: int, acc: pk.Acc[pk.float], viewA: pk.View1D[pk.float], viewB: pk.View2D[pk.float]):
+    acc += viewA[tid] * viewB[0][tid]
+
+def matmul(viewA, viewB):
+    """
+    1D Matrix Multiplication of compatible views
+
+    Parameters
+    ----------
+    viewA : pykokkos view
+    viewB : pykokkos view
+
+    Returns
+    -------
+    num : number
+    """
+    if not len(viewA.shape) == 1 or not viewA.shape[0] == viewB.shape[0]:
+        raise RuntimeError(
+            "Input operand 1 has a mismatch in its core dimension (Size {} is different from {})".format(viewA.shape[0], viewB.shape[0]))
+    
+    space = pk.ExecutionSpace.OpenMP
+
+    if str(viewA.dtype) == "DataType.double":
+        return pk.parallel_reduce(
+            pk.RangePolicy(space, 0, viewA.shape[0]),
+            matmul_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB)
+    
+    if str(viewA.dtype) == "DataType.float":
+        return pk.parallel_reduce(
+            pk.RangePolicy(space, 0, viewA.shape[0]),
+            matmul_impl_1d_float,
+            viewA=viewA,
+            viewB=viewB)
+
+@pk.workunit
+def divide_impl_1d_double(tid: int, view: pk.View1D[pk.double], viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double]):
+    view[tid] = viewA[tid] / viewB[tid]
+
+@pk.workunit
+def divide_impl_1d_float(tid: int, view: pk.View1D[pk.float], viewA: pk.View1D[pk.float], viewB: pk.View1D[pk.float]):
+    view[tid] = viewA[tid] / viewB[tid]
+
+
+def divide(viewA, viewB):
+    """
+    Divides positionally corresponding elements
+    of viewA with elements of viewB
+
+    Parameters
+    ----------
+    viewA : pykokkos view
+    viewB : pykokkos view
+
+    Returns
+    -------
+    view : pykokkos view
+    """
+    if str(viewA.dtype) == "DataType.double":
+        view = pk.View([viewA.shape[0]], pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            divide_impl_1d_double,
+            view=view,
+            viewA=viewA,
+            viewB=viewB)
+
+    elif str(viewA.dtype) == "DataType.float":
+        view = pk.View([viewA.shape[0]], pk.float)
+        pk.parallel_for(
+            viewA.shape[0],
+            divide_impl_1d_float,
+            view=view,
+            viewA=viewA,
+            viewB=viewB)
+
+    return view
+
+@pk.workunit
+def negative_impl_1d_double(tid: int, view: pk.View1D[pk.double], out: pk.View1D[pk.double]):
+    out[tid] = view[tid] * -1
+
+@pk.workunit
+def negative_impl_1d_float(tid: int, view: pk.View1D[pk.float], out: pk.View1D[pk.float]):
+    out[tid] = view[tid] * -1
+
+def negative(view):
+    """
+    Element-wise negative of the view
+
+    Parameters
+    ----------
+    view : pykokkos view
+
+    Returns
+    -------
+    view : pykokkos view
+    """
+    out = None
```

Comment:
So I followed the numpy's implementation for ufuncs and they always return the result in a new array. To make sure we use the same name everywhere, I renamed the output view in other funcs to "out" as well (it was previously called "view"). 

I also removed the `out = None` because it isnt needed.

Reply:
@tylerjereddy by default numpy allocates a new array  for all ufuncs right? But that isn't the approach we've been taking right? We've just been modifying the array in place?

Reply:
For `unary` ufuncs I think we've typically been modifying in place, yes. But most of the ufuncs here are `binary`, with two view operands, so it is unclear how you'd do anything other than provide a new view with the output. An `out` argument could be added later to control the location of the output result a bit more, but not sure we'd want to prioritize that just yet?

I certainly don't mind if we start providing a new view with the results for the `unary` cases as well--that's probably closer to what NumPy does, which in turn probably makes it easier for us to show performance improvements without changing the meaning in commonly-use algos.

Reply:
Yeah I think we do need an `out` argument eventually. I think that right now for the binary ufuncs it's fine to just allocate and return a new array.